test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,161 @@
+	deviceName = "exampledevice"
+	cdiVersion = "0.2.0"
+)
+
+type examplePlugin struct {
+	logger klog.Logger
+	d      plugin.DRAPlugin
+
+	cdiDir     string
+	driverName string
+}
+
+var _ drapbv1.NodeServer = &examplePlugin{}
+
+// runPlugin sets up the plugin and waits for connections from kubelet.
+func runPlugin(logger klog.Logger, cdiDir, driverName, endpoint, draAddress, pluginRegistrationPath string) error {
+	plugin := examplePlugin{
+		logger:     logger,
+		cdiDir:     cdiDir,
+		driverName: driverName,
+	}
+	if err := plugin.start(driverName, endpoint, draAddress, pluginRegistrationPath); err != nil {
+		return fmt.Errorf("start example plugin: %v", err)
+	}
+	// TODO: graceful termination via signal. Needs to be implemented in the caller.
+	select {}
+	plugin.stop()
+	return nil
+}
+
+// getJSONFilePath returns the absolute path where CDI file is/should be.
+func (ex *examplePlugin) getJSONFilePath(claimUID string) string {
+	return filepath.Join(ex.cdiDir, fmt.Sprintf("%s-%s.json", ex.driverName, claimUID))
+}
+
+// start sets up the servers that are necessary for a DRA kubelet plugin.
+func (ex *examplePlugin) start(driverName, endpoint, draAddress, pluginRegistrationPath string) error {
+	// Ensure that directories exist, creating them if necessary. We want
+	// to know early if there is a setup problem that would prevent
+	// creating those directories.
+	if err := os.MkdirAll(ex.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("create CDI directory: %v", err)
+	}
+	if err := os.MkdirAll(filepath.Dir(endpoint), 0750); err != nil {
+		return fmt.Errorf("create socket directory: %v", err)
+	}
+
+	d, err := kubeletplugin.Start(ex.logger, driverName, endpoint, draAddress, pluginRegistrationPath, ex)
+	if err != nil {
+		return fmt.Errorf("start kubelet plugin: %v", err)
+	}
+	ex.d = d
+
+	return nil
+}
+
+// stop ensures that all servers are stopped and resources freed.
+func (ex *examplePlugin) stop() {
+	ex.d.Stop()
+}
+
+// NodePrepareResource ensures that the CDI file for the claim exists. It uses
+// a deterministic name to simplify NodeUnprepareResource (no need to remember
+// or discover the name) and idempotency (when called again, the file simply
+// gets written again).
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	logger := klog.FromContext(ctx)
+	kind := ex.driverName + "/" + deviceName
+	filePath := ex.getJSONFilePath(req.ClaimUid)
+
+	// Determine environment variables.
+	var env map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &env); err != nil {
+		return nil, fmt.Errorf("unmarshal resource handle: %v", err)
+	}
+
+	// CDI wants env variables as set of strings.
+	envs := []string{}
+	for key, val := range env {
+		envs = append(envs, key+"="+val)
+	}
+
+	spec := specs.Spec{
+		Version: cdiVersion,
+		Kind:    kind,
+		Devices: []specs.Device{
+			{
+				Name: deviceName,
```

Comment:
deviceName should be unique for the same kind, I believe. We need to either generate it or just move Env into the top containerEdits and drop the device section.

Reply:
I don't think we can define a spec without any device.

Even this doesn't work:
```
	spec := specs.Spec{
		Version: "0.2.0",
		Kind:    kind,
		Devices: []specs.Device{
			{
				Name: deviceName,
			},
		},
		ContainerEdits: specs.ContainerEdits{
			Env: envs,
		},
	}
```

=> 

```
E0810 16:46:05.751933  138539 nonblockinggrpcserver.go:100] "dra: handling request failed" err="create spec: invalid CDI Spec: failed add device \"claim-0e3e71c0-97b4-447f-84f4-0ca7954cf76b\": invalid device, empty device edits" requestID=1 request="&NodePrepareResourceRequest{Namespace:default,ClaimUid:0e3e71c0-97b4-447f-84f4-0ca7954cf76b,ClaimName:test-inline-claim-resource,ResourceHandle:{\"user_a\":\"b\"},}"
```

Reply:
Right. I thought it should be possible to omit Devices/ContainerEdits at least. my bad, sorry.

Reply:
/lgtm

Reply:
> I thought it should be possible to omit Devices/ContainerEdits at least. my bad, sorry.

I agree that it would make sense semantically, but the current validation doesn't allow it and because keeping everything in one struct is shorter, I'm not going to ask for a change.